Agent throws SQL Exception when connection is closed by the server side

```
[Agent ERROR] - [scheduling-1] unknown or unreachable connection intercepted by agent! this data may not been recorded!
java.sql.SQLException: No operations allowed after statement closed.
	at com.mysql.cj.jdbc.exceptions.SQLError(SQLError:129)
	at com.mysql.cj.jdbc.exceptions.SQLError(SQLError:97)
	at com.mysql.cj.jdbc.exceptions.SQLError(SQLError:89)
	at com.mysql.cj.jdbc.exceptions.SQLError(SQLError:63)
	at com.mysql.cj.jdbc.exceptions.SQLError(SQLError:73)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:82)
	at com.mysql.cj.jdbc.StatementImpl.getConnection(StatementImpl.java:1357)
	at agent.plugin.mysql8.MySqlCounter.countExecutedNum(MySqlCounter.java:174)
	at agent.plugin.mysql8.MySqlCounter.add(MySqlCounter.java:132)
	at agent.plugin.mysql8.PreparedStatementHandler.after(PreparedStatementHandler.java:32)
	at com.mysql.cj.jdbc.ClientPreparedStatement.execute(ClientPreparedStatement.java)
	at com.alibaba.druid.filter.FilterChainImpl.preparedStatement_execute(FilterChainImpl.java:2931)
	at com.alibaba.druid.filter.FilterEventAdapter.preparedStatement_execute(FilterEventAdapter.java:440)
	at com.alibaba.druid.filter.FilterChainImpl.preparedStatement_execute(FilterChainImpl.java:2929)
	at com.alibaba.druid.filter.FilterEventAdapter.preparedStatement_execute(FilterEventAdapter.java:440)
	at com.alibaba.druid.filter.FilterChainImpl.preparedStatement_execute(FilterChainImpl.java:2929)
	at com.alibaba.druid.filter.FilterEventAdapter.preparedStatement_execute(FilterEventAdapter.java:440)
	at com.alibaba.druid.filter.FilterChainImpl.preparedStatement_execute(FilterChainImpl.java:2929)
	at com.alibaba.druid.filter.FilterAdapter.preparedStatement_execute(FilterAdapter.java:1058)
	at com.alibaba.druid.filter.FilterChainImpl.preparedStatement_execute(FilterChainImpl.java:2929)
	at com.alibaba.druid.proxy.jdbc.PreparedStatementProxyImpl.execute(PreparedStatementProxyImpl.java:131)
	at com.alibaba.druid.pool.DruidPooledPreparedStatement.execute$original$z6Vmt2ak(DruidPooledPreparedStatement.java:493)
	at com.alibaba.druid.pool.DruidPooledPreparedStatement.execute$original$z6Vmt2ak$accessor$FnSd81gJ(DruidPooledPreparedStatement.java)
	at com.alibaba.druid.pool.DruidPooledPreparedStatement$auxiliary$eyX08gKR.call(Unknown Source)
	at com.alibaba.druid.pool.DruidPooledPreparedStatement.execute(DruidPooledPreparedStatement.java)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:209)
	at org.jooq.impl.Tools.executeStatementAndGetFirstResultSet(Tools.java:3938)
	at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:294)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:368)
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:353)
	at org.jooq.impl.AbstractResultQuery.fetchInto(AbstractResultQuery.java:1549)
	at org.jooq.impl.SelectImpl.fetchInto(SelectImpl.java:3749)
	at com.keruyun.infra.ac.metrics.alert.db.dao.MetricsAlertObjectDao.getAlertListByTime(MetricsAlertObjectDao.java:34)
	at com.keruyun.infra.ac.metrics.alert.db.dao.MetricsAlertObjectDao$$FastClassBySpringCGLIB$$aaaabcd3.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:687)
	...
Caused by: com.mysql.cj.exceptions.StatementIsClosedException: No operations allowed after statement closed.
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:61)
	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:85)
	at com.mysql.cj.jdbc.StatementImpl.checkClosed(StatementImpl.java:340)
	at com.mysql.cj.jdbc.StatementImpl.getConnection(StatementImpl.java:1355)
	... 46 more
```

--- a/agent/agent-plugins/mysql-8.x/src/main/java/com/sbss/bithon/agent/plugin/mysql8/PreparedStatementInterceptor.java
+++ b/agent/agent-plugins/mysql-8.x/src/main/java/com/sbss/bithon/agent/plugin/mysql8/PreparedStatementInterceptor.java
@@ -18,6 +18,7 @@ package com.sbss.bithon.agent.plugin.mysql8;
 
 import com.sbss.bithon.agent.bootstrap.aop.AbstractInterceptor;
 import com.sbss.bithon.agent.bootstrap.aop.AopContext;
+import com.sbss.bithon.agent.bootstrap.aop.InterceptionDecision;
 import com.sbss.bithon.agent.core.metric.collector.MetricCollectorManager;
 import com.sbss.bithon.agent.core.metric.domain.sql.SqlCompositeMetric;
 import com.sbss.bithon.agent.core.metric.domain.sql.SqlMetricCollector;
@@ -41,12 +42,20 @@ public class PreparedStatementInterceptor extends AbstractInterceptor {
     }
 
     @Override
-    public void onMethodLeave(AopContext aopContext) throws Exception {
-        String methodName = aopContext.getMethod().getName();
+    public InterceptionDecision onMethodEnter(AopContext aopContext) throws Exception {
         Statement statement = (Statement) aopContext.getTarget();
-        String connectionString = MiscUtils.cleanupConnectionString(statement.getConnection()
+
+        // get the connection info before execution since the connection might be closed during execution
+        aopContext.setUserContext(MiscUtils.cleanupConnectionString(statement.getConnection()
                                                                              .getMetaData()
-                                                                             .getURL());
+                                                                             .getURL()));
+        return InterceptionDecision.CONTINUE;
+    }
+
+    @Override
+    public void onMethodLeave(AopContext aopContext) {
+        String methodName = aopContext.getMethod().getName();
+        String connectionString = aopContext.castUserContextAs();
 
         SqlCompositeMetric metric = sqlMetricCollector.getOrCreateMetric(connectionString);
         boolean isQuery = true;

--- a/agent/agent-plugins/mysql-8.x/src/main/java/com/sbss/bithon/agent/plugin/mysql8/StatementInterceptor.java
+++ b/agent/agent-plugins/mysql-8.x/src/main/java/com/sbss/bithon/agent/plugin/mysql8/StatementInterceptor.java
@@ -19,6 +19,7 @@ package com.sbss.bithon.agent.plugin.mysql8;
 
 import com.sbss.bithon.agent.bootstrap.aop.AbstractInterceptor;
 import com.sbss.bithon.agent.bootstrap.aop.AopContext;
+import com.sbss.bithon.agent.bootstrap.aop.InterceptionDecision;
 import com.sbss.bithon.agent.core.metric.collector.MetricCollectorManager;
 import com.sbss.bithon.agent.core.metric.domain.sql.SqlCompositeMetric;
 import com.sbss.bithon.agent.core.metric.domain.sql.SqlMetricCollector;
@@ -41,11 +42,19 @@ public class StatementInterceptor extends AbstractInterceptor {
     }
 
     @Override
-    public void onMethodLeave(AopContext aopContext) throws Exception {
+    public InterceptionDecision onMethodEnter(AopContext aopContext) throws Exception {
         Statement statement = (Statement) aopContext.getTarget();
-        String connectionString = MiscUtils.cleanupConnectionString(statement.getConnection()
+
+        // get the connection info before execution since the connection might be closed during execution
+        aopContext.setUserContext(MiscUtils.cleanupConnectionString(statement.getConnection()
                                                                              .getMetaData()
-                                                                             .getURL());
+                                                                             .getURL()));
+        return InterceptionDecision.CONTINUE;
+    }
+
+    @Override
+    public void onMethodLeave(AopContext aopContext) throws Exception {
+        String connectionString = aopContext.castUserContextAs();
 
         SqlCompositeMetric metric = sqlMetricCollector.getOrCreateMetric(connectionString);
         boolean isQuery = true;

